lock-impl/src/main/java/com/palantir/lock/impl/RateLimitingLockService.java

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright 2017 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.lock.impl;
+
+import java.math.BigInteger;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import com.palantir.lock.HeldLocksGrant;
+import com.palantir.lock.HeldLocksToken;
+import com.palantir.lock.LockClient;
+import com.palantir.lock.LockRefreshToken;
+import com.palantir.lock.LockRequest;
+import com.palantir.lock.LockResponse;
+import com.palantir.lock.LockServerOptions;
+import com.palantir.lock.LockService;
+import com.palantir.lock.SimpleHeldLocksToken;
+
+public class RateLimitingLockService implements LockService{
+    private static final ThreadCountLimiter globalLimiter = new ThreadCountLimiter(0);
+
+    private LockService delegate;
+    private ThreadCountLimiter limiter;
+
+    private RateLimitingLockService() {
+        // hidden
+    }
+
+    public static RateLimitingLockService create(LockService delegate, int maxServerThreads, int numClients) {
+        int threadPoolSize = maxServerThreads - 15;
```

Comment:
As mentioned offline, I'm concerned that the number of threads we need to "reserve" is a function of the number of clients, since each `LockServiceImpl` spins up two background threads. We should investigate this. Also this should be a constant.